docs/man/pydance.6
.TH pydance 6 "June 2nd, 2003"
.SH NAME
pydance \- a dancing simulator game
.SH SYNOPSIS
\fBpydance\fR [\fB\-\-filename \fIfilename \fR[ \fB\-\-mode\fI mode \fB\-\-difficulty\fI difficulty\fR ] ]
.SH OPTIONS
.IP \-\-help,\ \-h
Display a brief summary of command line options.
.IP \-\-version,\ \-v
Display the version of pydance (and features found, e.g. Psyco).
.IP \-\-filename\, \-f
Load and play a single file, then quit. By default, this is in
SINGLE (4 panel) mode, on BASIC difficulty.
.IP \-\-mode,\ \-m
Use a particular mode (e.g. VERSUS, 6PANEL, DMX). This only has an effect
when you also use \fB\-\-filename\fR.
.IP \-\-difficulty,\ \-d
Use a particular difficulty (e.g. TRICK, SMANIAC). This only has an effect
when you also use \fB\-\-filename\fR.
.SH DESCRIPTION
pydance is much like the popular arcade game "Dance Dance Revolution", in
which you stand on a mat on the floor, with 4 buttons pointing forward,
backwards, left, and right. Arrows scroll across the screen, and you must
step on the arrows on the mat when they reach the top. These steps
happen in time with music in the background, causing a rough approximation
of dancing.
.PP
However, pydance also plays more than DDR - it can simulate games like
Technomotion, Pop'n'Stage, Pump It Up, Para Para Paradise, and Dance ManiaX
(also called Dance Freaks).a
.PP
pydance supports most of the features of these games, such as background
and banner images for each song, multiple difficulty levels, freeze
arrows (arrows you have to remain standing on for a length of time), BPM
changes and stops, lyric displays, and so on. pydance also allows you to
use a real DDR mat if you have a Playstation to PC joystick adapter supported
by the ddrmat kernel module for Linux (which is a modified version of the
gamecon kernel module), or a Playstation to USB joystick adapter and the USB
joystick driver, for any OS. If you lack a mat, you can play on the
keyboard, but it's much less fun.
.PP
You can also make your own step files to new music, in pydance's \fB.dance\fR
format. \fIdance-spec.txt\fR included with pydance documents this format.
pydance can also read the popular Windows \fB.dwi\fR and \fB.sm\fR formats.
.SH USAGE
The first player's arrows and settings are controlled by the I, J, ,, and L
keys, and the second player's by the arrow keys. Enter starts a song or
activates a menu item.
.PP
pydance is documented more completely in the \fIREADME.html\fR file included
with it.
.SH FILES
.IP /usr/share/games/pydance/songs,\ /usr/local/share/games/pydance/songs,\ ~/.pydance/songs
By default, pydance will look in these directories for your songs, step files,
and other game data. You can configure this with the \fBsongdir\fR
option in your pydance.cfg file.
.IP
Note that for DWI files, you will need to put everything (the music file,
lyric file, background, and banner) in its own separate directory, and
that directory inside \fIanother\fR subdirectory in the \fIsongs\fR
folder. This is a limitation of the DWI format; it uses its path name
and files in the same directory, rather than specifying things in the
file itself like .dance files do.
.IP /usr/share/games/pydance/courses,\ /usr/local/share/games/pydance/courses,\ ~/.pydance/courses
Similar to above, but for course files. You can use \fBcoursedir\fR to
override this.
.IP /etc/pydance.cfg,\ ~/.pydance/pydance.cfg
pydance configuration files. The settings in your local one override
the system-wide settings.
.IP ~/.pydance/records
Your top scores for each song. This is a pickled Python file. It can be
safely deleted to reset your scores.
.IP ~/.pydance/input.cfg
Your input settings. This is a pickled Python file. It can be safely
deleted to reset your input configuration.
.IP /usr/share/games/pydance
Various data needed by pydance, such as background music and Python modules.
.SH AUTHORS
Brendan Becker <tgz@orotech.net>, Joe Wreschnig <piman@debian.org>, and Pavel Krivitsky <pavel@krivitsky.name>, with help from many others.
.PP
Many of the other dancing games mentioned here are trademarks of their
respective owners.
.SH SEE ALSO
\fBfindbpm\fR(1),
README included with pydance,
manual.html included with pydance,
dance-spec.txt included with pydance.


ChangeLog
dpydance ChangeLog

Legend:
+ New feature added or made an existing one better
- Removed a feature, or something bad happened
* Fixed a bug
! Changed something important, probably incompatibly.

----- - CURRENT
1.1.0 - Under new management.
+ A 2.5x speed multiplier was added.
+ A constant-BPM speed mode (from 200bpm up to 800bpm) was implemented.
+ Sorting/Floders by steps difficulty and rating added. (Frank Foeth)
+ UTF-8 BOMs now handled. (Stefan Kloe)
+ Pydance now has gettext support, with Catalan translation
  available. (Carles Pina i Estany and Catux LUG)
+ Non-ASCII song names are now printed to console if console encoding
  supports it.
+ Font theming support, including support for Asian song names,
  unsupported by pygame's default font.
- Song preview now starts at 0, to work around a pygame/libsdl
  bug. (Carles Pina i Estany)
* It is now guaranteed that songs in a course are not repeated.
* Fixes crackling sound caused by pygame mixer being initialized incorrectly.
* Minor typos fixed.
	
1.0.3 - Like that's possible.					[2005.04.04]
* An import conflict with python-stats was fixed.
+ Take a shot at detecting 3 axis 10 button joypads.
* Fix crash when using flat mode. (Drake Wilson)
* Better Xbox pad support. (David Fifield)
+ Change options between songs. (Pavel N. Krivitsky)
* Support for #FILE in DWI files. (Pavel N. Krivitsky)
* Fixes for player's best/worst courses. (Pavel N. Krivitsky)

1.0.2 - Moist!							[2004.07.06]
* Defafult announcer theme courtsey of DJ Nz17!
* Minor typos in the manual fixed.

1.0.1 -	blah blah BLAH blah blah BLAH bla			[2004.03.12]
* Tug of War lifebar no longer crashes.
* Standard make install target works correctly again.
* Fix a possible divide-by-zero related crash in the song selector.
* Avoid locale-setting error on Mac OS X.
* Fullscreen toggling on OS X and Windows works again.

1.0.0 - Oh, it's a snaaaaaaaaaaake! (on a dancepad!)		[2004.03.01]
+ The TODO file has been revised with plans for pydance 1.2.
+ Display a "Press Confirm" on the grading screen over the banner.
+ Add an icon (icon.png).
+ Using a pydance.zip (containing the modules needed) works now. (This
  is mostly of interest to packagers).
* Images added for all interface options.
* Fix a possible divide-by-zero in FPS calculuation.

0.9.1 - This release is dedicated to Gentoo bug #35890.		[2004.02.11]
+ Display the average time off (in milliseconds) when grading.
+ Page up and page down work again; they are back left and back right
  on the pad.
+ The manual now includes instructions for step and theme making, as
  well as a table of contents.
* Don't crash sometimes on slower systems when failing.
* Commas (or periods) are displayed on the grading screen in the score again.
* Fix a rare mixer-related crash.
* Load DWIs with subtitles in courses correctly again.
* Fix crash taking screenshots of the grading screen.
* Many smaller bugfixes from many people. Thanks, everyone.

0.9.0 - Great Three-Toed Sloth of Ice Planet Hoth!		[2004.02.02]
+ "Click only" assist tick.
+ New grading screen... with groove graph!
 + Thanks to sjbrown for the grade graphics.
+ Player's best/worst CRS file support.
+ There are now some command line options.
+ Player's best/worst and random courses generated internally.
 + Added "Player's Likes/Dislikes" for commonly or rarely played songs.
+ Changing BPMs are colored (thanks to xsdg).
+ A great attack lifebar was added (thanks to Jooon).
+ Lifebars are now animated.
+ Select (Random) behaves differently when viewing the folder list.
  (Thanks to Pavel).
+ A comprehensive new manual is available in docs/manual.html!
* Movies work again (thanks to Jon Åslund).
* A new "safe" preview mode that won't preview MP3s; this seems to stop
  most MP3-related crashes on Windows and SMP Unix systems. This is now
  the default mode on Windows.
! Start + Select now exits to the grading screen always, by popular demand.
! .step support removed.
! All your records will be deleted upon running this.
! pydance no longer searches "pyddr" folders by default.

0.8.4 - We must force those people to date actual gamer nerds.	[2004.01.08]
+ EZ2 and EZ2 Real support.
* Fix DWI parsing bugs and typos introduced in 0.8.3.
* Screenshot support is back, using Print Screen (or s on the grading screen.)
* Fix hold arrow + random transform bug.

0.8.3 - Also it should be called Streetball.			[2004.01.04]
!!! Completely new user interface for almost all screens. Please pay
  attention to the help messages displayed onscreen, because many of the
  controls have changed.
  + CD title and advanced BPM display support.
+ 3 panel mode (automatically generated from 6 panel beginner steps in
  .dwi, or available directly in .dance).
+ Hold arrows now have a grace period (by default 1/4 second, but it's
  affected by the judge type and difficulty).
+ Best grades for songs are now stored (in ~/.pydance/records).
+ Arrow animations have been tweaked to make songs with lots of 24th notes
  (like bag) easier to read.
+ Rewrite of findbpm by Paul Crowley to do least-squares estimation.
+ Non-EMSUSB2 adapters work again.

0.8.2 - I"M GONNA EAT YOU!!!1					[2003.11.16]
+ Input keys/buttons can now be remapped in the game.
 + Default keys are the numeric pad for player 1 and uiojklnm, for player 2.
 + Arrow keys, if unmapped, work for player 1 as well.
 ! The 'keyboard' option is gone.
 ! The 'saveinput' option saves an input.cfg file (in . or ~/.pydance) on
   exit. Delete this file to reset your mappings to defaults.
+ Animated arrows in the 'ddresque', 'vividesque', and 'soloesque' themes.
 - This release may be slower, especially for animated arrows.
+ Courses (Oni/Nonstop) support; put the files in ~/.pydance/courses
  courses/ under your pydance installation directory. Currently only
  DWI/SM-style .crs files are supported.
! The 'mrt' announcer theme is now distributed separately from the main source.
! Support for MAD/AO audio output was removed. It didn't work reliably anyway.

pydance.py
#!/usr/bin/env python
# pydance - a dancing game written in Python

import os
import sys
from getopt import getopt

VERSION = "1.1.0"

from i18n import *

# fuck you, Python.
def print_help():
  print
  print _("Usage: %s [options]") % sys.argv[0]
  print _(" -h, --help         display this help text and exit")
  print _(" -v, --version      display the version and exit")
  print _(" -f, --filename     load and play a step file")
  print _(" -m, --mode         the mode to play the file in (default SINGLE)")
  print _(" -d, --difficulty   the difficult to play the file (default BASIC)")
  raise SystemExit

def print_version():
  print _("pydance %s by Joe Wreschnig, Brendan Becker, and others") % VERSION
  print "pyddr-discuss@icculus.org - http://icculus.org/pyddr"
  raise SystemExit

if len(sys.argv) < 2: pass 
elif sys.argv[1] in ["--help", "-h"]: print_help()
elif sys.argv[1] in ["--version", "-v"]: print_version()

# Don't import anything that initializes the joysticks or config until
# after we're (reasonably) sure no one wants --help or --version.
from constants import * # This needs to be here to set sys.path.

import util
import games
import dance
import pygame
import courses
import colors
import records
import menudriver

from fileparsers import SongItem
from pygame.mixer import music
from fontfx import TextProgress
from error import ErrorMessage
from fonttheme import FontTheme

from pad import pad

# Set our required display paramters. Currently, this is nothing
# strange on any platforms, but in the past and likely in the future
# some platforms need other flags.
def set_display_mode():
  try:
    flags = 0
    if mainconfig["fullscreen"]: flags |= FULLSCREEN
    screen = pygame.display.set_mode([640, 480], flags, 16)
  except:
    raise SystemExit(_("E: Can't get a 16 bit display! pydance doesn't work in 8 bit mode."))
  return screen

# Load a single song (given the filename) and then play it on the
# given difficulty.
def play_and_quit(fn, mode, difficulty):
  print _("Entering debug (play-and-quit) mode.")
  screen = set_display_mode()  
  pygame.display.set_caption("pydance " + VERSION)
  pygame.mouse.set_visible(0)
  pc = games.GAMES[mode].players
  dance.play(screen, [(fn, [difficulty] * pc)],
             [player_config] * pc, game_config, mode)
  raise SystemExit

# Pass a list of files to a constructor (Ctr) that takes the filename
# as the first argument, and the args tuple as the rest.
def load_files(screen, files, type, Ctr, args):
  if len(files) == 0: return []

  screen.fill(colors.BLACK)
  pct = 0
  inc = 100.0 / len(files)
  # Remove duplicates
  files = list(dict(map(None, files, [])).keys())
  objects = []
  message = _("Found %d %s. Loading...") % (len(files), _(type))
  pbar = TextProgress(FontTheme.loading_screen, message, colors.WHITE, colors.BLACK)
  r = pbar.render(0).get_rect()
  r.center = [320, 240]
  for f in files:
    try: objects.append(Ctr(*((f,) + args)))
    except RuntimeError, message:
      print _("E:"), f
      print _("E:"), message
      print
    except Exception, message:
      print _("E: Unknown error loading"), f
      print _("E:"), message
      print _("E: Please contact the developers (pyddr-devel@icculus.org).")
      print
    pct += inc
    img = pbar.render(pct)
    pygame.display.update(screen.blit(img, r))

  return objects

# Support fullscreen on Win32 / OS X?
if osname != "posix": pygame.display.toggle_fullscreen = set_display_mode
else: pass

# Actually start the program running.
def main():
  print "pydance", VERSION, "<pyddr-discuss@icculus.org> - irc.freenode.net/#pyddr"

  if mainconfig["usepsyco"]:
    try:
      import psyco
      print _("Psyco optimizing compiler found. Using psyco.full().")
      psyco.full()
    except ImportError: print _("W: Psyco optimizing compiler not found.")

  # default settings for play_and_quit.
  mode = "SINGLE"
  difficulty = "BASIC"
  test_file = None
  for opt, arg in getopt(sys.argv[1:],
                         "hvf:d:m:", ["help", "version", "filename=",
                                      "difficulty=", "mode="])[0]:
    if opt in ["-h", _("--help")]: print_help()
    elif opt in ["-v", _("--version")]: print_version()
    elif opt in ["-f", _("--filename")]: test_file = arg
    elif opt in ["-m", _("--mode")]: mode = arg
    elif opt in ["-d", _("--difficulty")]: difficulty = arg

  if test_file: play_and_quit(test_file, mode, difficulty)

  song_list = []
  course_list = []
  for dir in mainconfig["songdir"].split(os.pathsep):
    print _("Searching for songs in"), dir
    song_list.extend(util.find(dir, ['*.dance', '*.dwi', '*.sm', '*/song.*']))
  for dir in mainconfig["coursedir"].split(os.pathsep):
    print _("Searching for courses in"), dir
    course_list.extend(util.find(dir, ['*.crs']))

  screen = set_display_mode()
  
  pygame.display.set_caption("pydance " + VERSION)
  pygame.mouse.set_visible(False)
  try:
    if os.path.exists("/usr/share/pixmaps/pydance.png"):
      icon = pygame.image.load("/usr/share/pixmaps/pydance.png")
    else: icon = pygame.image.load(os.path.join(pydance_path, "icon.png"))
    pygame.display.set_icon(icon)
  except: pass

  music.load(os.path.join(sound_path, "menu.ogg"))
  music.play(4, 0.0)

  songs = load_files(screen, song_list, _("songs"), SongItem, (False,))

  # Construct the song and record dictionaries for courses. These are
  # necessary because courses identify songs by title and mix, rather
  # than filename. The recordkey dictionary is needed for player's
  # picks courses.
  song_dict = {}
  record_dict = {}
  for song in songs:
    mix = song.info["mix"].lower()
    title = song.info["title"].lower()
    if song.info["subtitle"]: title += " " + song.info["subtitle"].lower()
    if not song_dict.has_key(mix): song_dict[mix] = {}
    song_dict[mix][title] = song
    record_dict[song.info["recordkey"]] = song

  crs = load_files(screen, course_list, _("courses"), courses.CourseFile,
                   (song_dict, record_dict))
  crs.extend(courses.make_players(song_dict, record_dict))
  records.verify(record_dict)

  # Let the GC clean these up if it needs to.
  song_list = None
  course_list = None
  record_dict = None
  pad.empty()

  if len(songs) < 1:
    ErrorMessage(screen,
                 (_("You don't have any songs or step files. Check out "
                  "http://icculus.org/pyddr/get.php#songs "
                  "and download some free ones. "
                  "If you already have some, make sure they're in ")) +
                 mainconfig["songdir"])
    raise SystemExit(_("You don't have any songs. Check http://icculus.org/pyddr/get.php#songs ."))

  menudriver.do(screen, (songs, crs, screen))

  # Clean up shit.
  music.stop()
  pygame.quit()
  mainconfig.write(os.path.join(rc_path, "pydance.cfg"))
  # FIXME -- is this option a good idea?
  if mainconfig["saveinput"]: pad.write(os.path.join(rc_path, "input.cfg"))
  records.write()

if __name__ == '__main__': main()
